### ISSUE TITLE
No Cookies during precache/cache request

### ISSUE THREAD

**Issue by chenav** [GoogleChrome/sw-precache#62]
Hey,

The sw-precache does not seem to allow the fetch feature, including cookies / credentials.
Actually, Grunt generates a default fetch, which does not send cookies/credentials.
For instance, it is quite problematic when you want to precache dynamic resources from an app server.

Fetch API allows it, with mode and credentials attributes.

I may be wrong, but I did not find it in the precache API.

C.

**Comment by jeffposnick** [comment]
Can you tell me a bit more about your use case? The way I envision folks using `sw-precache`, the assets are are fetched immediately when the service worker is registered, which (usually) would happen upon first visit to a web app. Are you delaying service worker registration until after a user has logged in to your web app?

I've got some concerns around having `sw-precache` maintain a cache of assets that are personalized to a given user based on some sort of user credential. One problem is that you'd need to properly expire all those personalized assets whenever the user logs out, and I don't know if that could be implemented generically in a library like `sw-precache`. It's the sort of thing that is (probably) best handled via service worker code customized for your situation.

This [Stack Overflow question/answer](http://stackoverflow.com/questions/34035667/service-worker-making-extra-call) has some further discussion about how authenticated requests complicate service worker caching.

**Comment by jeffposnick** [comment]
...and actually, on further inspection of the code, it's already setting [`{credentials: 'same-origin'}`](https://github.com/GoogleChrome/sw-precache/blob/master/service-worker.tmpl#L77) on the precaching request. Are you asking for remote origin credentials, specifically?

**Comment by chenav** [comment]
Hey Jeffrey,

The use case is the following:

- I have a CMS which delivers dozens of sites (web apps actually), with a unique hash for each in the URL
- the front end is driven by an Angular app
- the app has an app shell (minified js + css + html) and REST calls for the contents
- I am testing service workers to deliver cache optimization, push, and offline in the near future

1. Through grunt, I prepared the precache of static files (the app shell, common for all sites my CMS delivers), just as described in the tutorial, within the writeServiceWorkerFile function. 
Works fine.

2. I wanted to try a full offline scenario (app shell + REST dynamic content).
Then in a « custom-dynamic-rules.js » file, that I generate on-the-fly per site, I prepare the tree/list of additional dynamic URLs to precache (they are specific per sites).
For that, I used the function:

toolbox.precache([« /srv/service/mysuperhomepage?hash=12345 » ],[« /srv/service/mysupercontent?id=12],… );

And of course, within my gruntfile, I inserted that file through the sw-precache attribute command of the writeServiceWorkerFile function: 
importScripts: [ ‘/path/sw-toolbox.js',  ’/srv/custom-dynamic-rules.js’]

3. I get in the service-worker.js file generated by grunt:
...
importScripts("./js/sw-toolbox.js", »./srv/custom-dynamic-rules.js »);

var PrecacheConfig = [[« /appshell/filea.html »,"checksum1 » ], [« /appshell/fileb.js», « checksum2 » ],…] ;
…

In that scenario, the toolbox.precache function (from the imported custom-dynamic-rules.js file) does  HTTP GET requests with a cookie-set A.
And the precacheConfig table containing my app shell files generates a bunch of HTTP GET with a cookie-set B.
In terms of analytics and/or credentials, it was a problem. 

I did some tests after I read you message: If I insert my ’/srv/custom-dynamic-rules.js’ in the writeServiceWorkerFile of the grunt, it actually generates a HTTP GET with the same cookie-set as the other app shell files. It really works as you mentionned.

But as my app shell is always the same for all my sites, and my custom-dyanmic-rules are different for each site: I tried to separate the sw precache calls.
I realize now that the toolbox.precache command does not share the same context as the grunt precache command (at least, the way I did it).

I will rethink my process, now I know that.

Thanks for your answer !

C.
